### PR TITLE
msg validation check for `/initialpose`  in `nav2_amcl`

### PR DESCRIPTION
**Pull request by GoesM**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #4294  |
| Primary OS tested on | Ubuntu22.04 |
| Robotic platform tested on |  gazebo simulation of Tally |
| Does this PR contain AI generated software? | No |

---

## Description of contribution in a few bullet points
- make a validation util  for `array` 
- make a validation util for `PoseWithCovariance`
- make a validation util for `PoseWithCovarianceStamped`
- add test for new changes
- use the validation check for `/initialpose` in `nav2_amcl`
---

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

File: nav2_util/include/nav2_util/validate_messages.hpp [hunk 1]
```diff
@@ -55,6 +55,17 @@ bool validateMsg(const double & num)
   return true;
 }
 
+template<size_t N>
+bool validateMsg(const std::array<double, N> & msg)
```
Comment (SteveMacenski) [inline nav2_util/include/nav2_util/validate_messages.hpp:59]: Vector?

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

File: nav2_util/include/nav2_util/validate_messages.hpp [hunk 2]
```diff
@@ -111,6 +122,23 @@ bool validateMsg(const geometry_msgs::msg::Pose & msg)
...
+bool validateMsg(const geometry_msgs::msg::PoseWithCovariance & msg)
+{
+  // check sub-type
+  if (!validateMsg(msg.pose)) {return false;}
+  if (!validateMsg(msg.covariance)) {return false;}
+
+  return true;
+}
+
+bool validateMsg(const geometry_msgs::PoseWithCovarianceStamped_ & msg)
```
[the remark this answers is not in the document]
Comment (SteveMacenski) [reply nav2_util/include/nav2_util/validate_messages.hpp:134]: You may need to include the extra header

File: nav2_util/include/nav2_util/validate_messages.hpp (hunk 1 above)
Comment (GoesM) [reply nav2_util/include/nav2_util/validate_messages.hpp:59]: I noticed `YAML::List` is always transformed into array in ROS, which is defined in `geometry_msg`:

```cpp
  using _pose_type =
    geometry_msgs::msg::Pose_<ContainerAllocator>;
  _pose_type pose;
  using _covariance_type =
    std::array<double, 36>;
  _covariance_type covariance;
```

**Comment by GoesM** [comment]
uh.... the error is odd to me 
```error: ‘PoseWithCovarianceStamped’ in namespace ‘geometry_msgs::msg’ does not name a type```

My local test (based on ros2-humble) run successfully however failed during CI.

Is the type name different between `main` and `humble` ?

**Comment by SteveMacenski** [comment]
I think you're just missing `#include "geometry_msgs/msg/<messsage>.hpp"` for the message

File: nav2_util/include/nav2_util/validate_messages.hpp (hunk 1 above)
Comment (SteveMacenski) [reply nav2_util/include/nav2_util/validate_messages.hpp:59]: Ah got it, this is for the covariance checking :+1: Can you add a comment for this one that this is what this is for? I'll forget in a year and when I look over this again I'll wonder again why its not a vector :laughing:

Comment (GoesM) [reply nav2_util/include/nav2_util/validate_messages.hpp:59]: of course 😆

**Comment by GoesM** [comment]
You're right! ^_^

though I'm still confused why it could be built in `humble` fork ( within my local environment ), T_T, it seems that it doesn't affect much.

> I think you're just missing `#include "geometry_msgs/msg/<messsage>.hpp"` for the message

After all, everything seems fine after changes following your reminder. ^_^

**Review by SteveMacenski — APPROVED** [review]

**Pull request merged**